src/witan/send/adroddiad/ncy.clj

```diff
@@ -1,7 +1,298 @@
+                           (tc/first)
+                           (tc/ungroup)
+                           (tc/rename-columns {cy  :cy-ref
+                                               ncy :ncy-ref})
+                           (tc/set-dataset-name "_ref"))
+                       [id])
+         (tc/map-columns ncy-imputation [ncy cy :ncy-ref :cy-ref]
+                         (fn [ncy cy ncy-ref cy-ref]
+                           (when (and (nil? ncy) (some? cy) (some? ncy-ref) (some? cy-ref))
+                             (str "Imputed from NCY=" ncy-ref " in " cy-ref ))))
+         (tc/map-columns ncy-imputed    [ncy cy :ncy-ref :cy-ref ncy-imputation]
+                         (fn [ncy cy ncy-ref cy-ref ncy-imputation]
+                           (if (and (nil? ncy) (some? ncy-imputation))
+                             (- ncy-ref (- cy-ref cy))
+                             ncy)))
+         (tc/drop-columns #(= "_ref" %) :src-table-name)))))
+
+
+(rcf/tests                              ; Tests for impute-nil-ncy
+
+ "Where all :academic-year are nil (so no reference rows), should just add nil :academic-year column."
+ (-> (tc/dataset {:id [-1 -0] :calendar-year 2000 :academic-year nil})
+     (impute-nil-ncy)) :=
+ (-> (tc/dataset {:id [-1 -0] :calendar-year 2000 :academic-year nil :academic-year-imputation nil}))
+
+
+ "For a subject with a reference row (`:id`s 2 to 5 in test data below),
+  should impute NCY for any rows with nil NCY, otherwise leaves as is.
+   NOTE: NCY may be imputed outside the SEND range of -4 to 20.
+   NOTE: Ignoring :academic-year-imputation column for the moment"
+ (-> (tc/concat (tc/dataset {:id  1 :calendar-year [2000 2001 2002] :academic-year      nil     })
+                (tc/dataset {:id  2 :calendar-year [2000 2001 2002] :academic-year [nil   1 nil]})
+                (tc/dataset {:id  3 :calendar-year [2000 2001 2002] :academic-year [nil nil   2]})
+                (tc/dataset {:id  4 :calendar-year [2000 2001 2002] :academic-year [nil nil  -4]})
+                (tc/dataset {:id  5 :calendar-year [2000 2001 2002] :academic-year [ 20 nil nil]})
+                (tc/dataset {:id  6 :calendar-year [2000 2001 2002] :academic-year [  0   1   2]}))
+     (impute-nil-ncy)
+     (tc/drop-columns :academic-year-imputation)
+     (tc/order-by [:id :calendar-year])) :=
+ (-> (tc/concat (tc/dataset {:id  1 :calendar-year [2000 2001 2002] :academic-year      nil     })
+                (tc/dataset {:id  2 :calendar-year [2000 2001 2002] :academic-year [  0   1   2]})
+                (tc/dataset {:id  3 :calendar-year [2000 2001 2002] :academic-year [  0   1   2]})
+                (tc/dataset {:id  4 :calendar-year [2000 2001 2002] :academic-year [ -6  -5  -4]})
+                (tc/dataset {:id  5 :calendar-year [2000 2001 2002] :academic-year [ 20  21  22]})
+                (tc/dataset {:id  6 :calendar-year [2000 2001 2002] :academic-year [  0   1   2]})))
```

Comment:
feels like you are setting up a bit of test data and then checking a number of different cases that could fail in different ways. Worth splitting into separate tests on the same dataset?